components/datetime/src/time_zone.rs

```diff
@@ -417,282 +740,457 @@ impl TimeZoneFormat {
+            Self::SpecificNonLocationLong(unit) => unit.format(sink, time_zone, data_payloads),
+            Self::SpecificNonLocationShort(unit) => unit.format(sink, time_zone, data_payloads),
+            Self::GenericLocation(unit) => unit.format(sink, time_zone, data_payloads),
+            Self::LocalizedGmt(unit) => unit.format(sink, time_zone, data_payloads),
+            Self::Iso8601(unit) => unit.format(sink, time_zone, data_payloads),
+            Self::ExemplarCity(unit) => unit.format(sink, time_zone, data_payloads),
+        }
     }
+}
 
+impl FormatTimeZone for GenericNonLocationLongFormat {
     /// Writes the time zone in long generic non-location format as defined by the UTS-35 spec.
     /// e.g. Pacific Time
     /// https://unicode.org/reports/tr35/tr35-dates.html#Time_Zone_Format_Terminology
-    pub(super) fn long_generic_non_location_format<W: fmt::Write + ?Sized>(
+    fn format<W: fmt::Write + ?Sized>(
         &self,
         sink: &mut W,
         time_zone: &impl TimeZoneInput,
-    ) -> Result<(), DateTimeFormatError> {
-        sink.write_str(
-            self.mz_generic_long
-                .as_ref()
-                .map(|p| p.get())
-                .and_then(|metazones| {
-                    time_zone
-                        .time_zone_id()
-                        .and_then(|tz| metazones.overrides.get(tz))
-                })
-                .or_else(|| {
-                    self.mz_generic_long
-                        .as_ref()
-                        .map(|p| p.get())
-                        .and_then(|metazones| {
-                            time_zone
-                                .metazone_id()
-                                .and_then(|mz| metazones.defaults.get(mz))
-                        })
-                })
-                .ok_or(fmt::Error)?,
-        )
-        .map_err(DateTimeFormatError::from)
+        data_payloads: &TimeZoneDataPayloads,
+    ) -> Result<fmt::Result, DateTimeFormatError> {
+        let formatted_time_zone: Option<&str> = data_payloads
+            .mz_generic_long
+            .as_ref()
+            .map(|p| p.get())
+            .and_then(|metazones| {
+                time_zone
+                    .time_zone_id()
+                    .and_then(|tz| metazones.overrides.get(tz))
+            })
+            .or_else(|| {
+                data_payloads
+                    .mz_generic_long
+                    .as_ref()
+                    .map(|p| p.get())
+                    .and_then(|metazones| {
+                        time_zone
+                            .metazone_id()
+                            .and_then(|mz| metazones.defaults.get(mz))
+                    })
+            });
+
+        match formatted_time_zone {
+            Some(ftz) => match sink.write_str(ftz) {
+                Ok(_) => Ok(Ok(())),
+                Err(_) => Ok(Err(fmt::Error)),
+            },
```

Comment:
Here and elsewhere: Suggestion to simplify the code:

```suggestion
            Some(ftz) => Ok(match sink.write_str(ftz)),
```

Reply:
hmm, looks like it should be `Some(ftz) => Ok(sink.write_str(ftz)),`